### PR TITLE
Zpi 67 backend tests

### PR DESCRIPTION
**Pull request by Nightbaka**
Added unit tests. Modified the cluster_crud to be testable

### CODE REVIEW

File: backend/cluster/cluster_crud_test.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	mockGetResourceI := new(mock.Mock)
+	mockGetResourceI.On("func1", "Pod", "default", "default").Return(&MockResourceInterface{}, &models.ModelError{Code: 404, Message: "Not found"})
+	getResourceI = func(resourceType string, namespace string, emptyNamespace string) (dynamic.ResourceInterface, *models.ModelError) {
+		args := mockGetResourceI.Called(resourceType, namespace, emptyNamespace)
+		return args.Get(0).(dynamic.ResourceInterface), args.Get(1).(*models.ModelError)
+	}
+
+	t.Run("Test GetResource", func(t *testing.T) {
+		_, err := GetResource("Pod", "default", "default")
+		assert.NotNil(t, err)
```
Comment (samuelus) [inline backend/cluster/cluster_crud_test.go:39]: Change it to verify if received error is exactly what we want - message, code, and is this an object that we want, not sth else

Comment (samuelus) [reply backend/cluster/cluster_crud_test.go:39]: General comment

File: backend/cluster/cluster_crud_test.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+		expectedResourceDetails := &models.ResourceDetails{
+			ResourceDetails: func() *interface{} {
+				obj := &unstructured.Unstructured{
+					Object: map[string]interface{}{
+						"key": "value",
+					},
+				}
+				var i interface{} = obj
+				return &i
+			}(),
```
Comment (samuelus) [inline backend/cluster/cluster_crud_test.go:70]: does it need to be implemented in this way?

**Review by samuelus — CHANGES_REQUESTED** [review]

File: backend/cluster/cluster_crud_test.go (hunk 2 above)
Comment (Nightbaka) [reply backend/cluster/cluster_crud_test.go:70]: Pretty sure yes as our ResourceDetails model has *interface not interface type. I will extract the mock resource creation into a separate function so the test code can be more readable
image

File: backend/cluster/cluster_crud.go [hunk 3]
```diff
@@ -9,8 +9,10 @@ import (
 
 const DefaultNamespace = "default"
 
+var getResourceI = getResourceInterface
```
[the remark this answers is not in the document]
Comment (Nightbaka) [reply backend/cluster/cluster_crud.go]: Changed, now function is one of the arguments to crud functions

File: backend/cluster/cluster_crud_test.go (hunk 2 above)
Comment (Nightbaka) [reply backend/cluster/cluster_crud_test.go]: Put it in a separate helper function with logic that is maybe more readable

File: backend/cluster/cluster_crud_test.go (hunk 1 above)
Comment (Nightbaka) [reply backend/cluster/cluster_crud_test.go:78]: resolved, I should be checking all error values now

**Comment by Nightbaka** [comment]
I will add the tests for listing resources as separate PR as they will propably have a different scope. This one is ready for merge, includes unit tests for all C R U D on resources

**Review by samuelus — APPROVED** [review]

**Pull request merged**